### DISCUSSION TITLE
creating-a-source-generator-part-5-finding-a-type-declarations-namespace-and-type-hierarchy/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#90] [Announcements]
# creating-a-source-generator-part-5-finding-a-type-declarations-namespace-and-type-hierarchy/

In this post I show a couple of handy patterns for source generators: finding a type declaration's namespace and its class hierarchy

https://andrewlock.net/creating-a-source-generator-part-5-finding-a-type-declarations-namespace-and-type-hierarchy/

**Comment by yoelhalb-hitech** [comment]
There are is a bug in the code:

		while (true)
		{
			if (namespaceParent.Parent is not NamespaceDeclarationSyntax parent)
			{
				break;
			}

			// Add the outer namespace as a prefix to the final namespace
			nameSpace = $"{namespaceParent.Name}.{nameSpace}"; // This will give the child namespace twice
			namespaceParent = parent;
		}

it should be instead:

		while (true)
		{
			if (namespaceParent.Parent is not NamespaceDeclarationSyntax parent)
			{
				break;
			}

			// Add the outer namespace as a prefix to the final namespace			
			namespaceParent = parent; // Set first the namepaceParent to parent
            nameSpace = $"{namespaceParent.Name}.{nameSpace}"; // Should now work correctly
		}

[a comment by an automated account is not shown]

**Comment by psulek** [comment]
Isn't there any nuget package with such helper methods for extracting namespace, parent class hierarchy,etc? So we can focus on writing our source generator and not reinventing the wheel? It will help a lot for developers who's starting with source code generators.

**Reply by yoelhalb-hitech** [reply]
I actually wrote some utility extensions on [GitHub](https://github.com/yoelhalb-hitech/DotNetPowerExtensions/tree/main/DotNetPowerExtensions.RoslynExtensions) and it is on Nuget as [SequelPay.DotNetPowerExtensions.RoslynExtensions](https://www.nuget.org/packages/SequelPay.DotNetPowerExtensions.RoslynExtensions/)

It is a shame though that these aren't part of the core libraries

**Reply by psulek** [reply]
@yoelhalb-hitech Nice thanks, will look at it!
